polkadot/xcm/xcm-executor/src/lib.rs

```diff
@@ -633,10 +646,26 @@ impl<Config: config::Config> XcmExecutor<Config> {
 				let assets = Self::reanchored(deposited, &dest, None);
 				let mut message = vec![ReserveAssetDeposited(assets), ClearOrigin];
 				message.extend(xcm.0.into_iter());
+
+				// put back parked_fee in holding register to be charged by XcmSender
+				self.holding.subsume_assets(parked_fee);
 				self.send(dest, Xcm(message), FeeReason::DepositReserveAsset)?;
 				Ok(())
 			},
 			InitiateReserveWithdraw { assets, reserve, xcm } => {
+				// we need to do this take/put cycle to solve wildcards and get exact assets to be
+				// weighed
+				let to_weigh = self.holding.saturating_take(assets.clone());
+				self.holding.subsume_assets(to_weigh.clone());
+
+				let mut message_to_weigh =
+					vec![WithdrawAsset(to_weigh.into()), ClearOrigin];
+				message_to_weigh.extend(xcm.0.clone().into_iter());
+				let (_, fee) = validate_send::<Config::XcmSender>(reserve, Xcm(message_to_weigh))?;
```

Comment:
As I understand fee is calculated in the native asset (for example, KSM). How will it work for the cases when, for example, RMRK or USTD is transferred from the Kusama Asset Hub and no KSM in the holding? Will the extrinsic itself fail on the Asset Hub and the state will be rolled back?

Reply:
Until fees can be paid in another asset, there should always be enough KSM to pay for fees, not only for delivery, also for execution.

Reply:
Does this mean that KSM must be included into the xcm together with the sending token (USDT)?

Reply:
Yes. It should've always been the case in order to pay for the regular execution fees.